cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/ETLSpark.java

```diff
@@ -0,0 +1,179 @@
+      setExecutorResources(config.getResources());
+    }
+
+    // add source, sink, transform ids to the properties. These are needed at runtime to instantiate the plugins
+    Map<String, String> properties = new HashMap<>();
+    properties.put(Constants.PIPELINEID, GSON.toJson(pipelineIds));
+    setProperties(properties);
+  }
+
+  @Override
+  public void beforeSubmit(SparkContext context) throws Exception {
+    cleanupFiles = new ArrayList<>();
+    finishers = new ArrayList<>();
+
+    Map<String, String> properties = context.getSpecification().getProperties();
+    Pipeline pipeline = GSON.fromJson(properties.get(Constants.PIPELINEID), Pipeline.class);
+    String sourcePluginId = pipeline.getSource();
+
+    PluginContext pluginContext = context.getPluginContext();
+    BatchConfigurable<BatchSourceContext> batchSource = pluginContext.newPluginInstance(sourcePluginId);
+    SparkBatchSourceContext sourceContext = new SparkBatchSourceContext(context,
+                                                                        new DatasetContextLookupProvider(context),
+                                                                        sourcePluginId);
+    batchSource.prepareRun(sourceContext);
+
+    SparkBatchSourceFactory sourceFactory = sourceContext.getSourceFactory();
+    if (sourceFactory == null) {
+      // TODO: Revisit what exception to throw
+      throw new IllegalArgumentException("Input is not set");
+    }
+    addFinisher(batchSource, sourceContext, finishers);
+
+    // setup time partition for each error dataset
+//    for (TransformInfo transformInfo : transformInfos) {
+//      if (transformInfo.getErrorDatasetName() != null) {
+//        addErrorDataset(transformInfo.getErrorDatasetName(), context);
+//      }
+//    }
+
+    SparkBatchSinkFactory sinkFactory = new SparkBatchSinkFactory();
+    List<SinkInfo> sinkInfos = pipeline.getSinks();
+    for (SinkInfo sinkInfo : sinkInfos) {
+      BatchConfigurable<BatchSinkContext> batchSink = pluginContext.newPluginInstance(sinkInfo.getSinkId());
+      BatchSinkContext sinkContext = new SparkBatchSinkContext(sinkFactory, context, null, sinkInfo.getSinkId());
+      batchSink.prepareRun(sinkContext);
+      addFinisher(batchSink, sinkContext, finishers);
+//      if (sinkInfo.getErrorDatasetName() != null) {
+//        addPropertiesToErrorDataset(sinkInfo.getErrorDatasetName(), context);
+//      }
+    }
+
+    File configFile = File.createTempFile("ETLSpark", ".config");
+    cleanupFiles.add(configFile);
+    try (OutputStream os = new FileOutputStream(configFile)) {
+      sourceFactory.serialize(os);
+      sinkFactory.serialize(os);
+    }
+
+    context.localize("ETLSpark.config", configFile.toURI());
+  }
+
+  @Override
+  public void onFinish(boolean succeeded, SparkContext context) throws Exception {
+    for (Finisher finisher : finishers) {
+      finisher.onFinish(succeeded);
+    }
+    for (File file : cleanupFiles) {
+      file.delete();
```

Comment:
If the first file throws an exception, the rest of the files won't be cleaned up. perhaps it is better to do:

```java 
for (File file : cleanupFiles){
    try { 
      file.delete();
    } catch (IOException exception ) {
       LOG.error("Failed to clean up resource {} ", file, exception);
    }
 }